### PR TITLE
Adding common reference specs to be used while designing and testing

### PR DESCRIPTION
**Pull request by Arafatk**
This Pr adds a list of test files that includes a set of basic design specs to be satisfied by the newly added plots.  The API must match these as much as possible.

### CODE REVIEW

File: spec/common_scripting_layer/line_spec.rb [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    setup_data
+    plot = Rubyplot::Line.new(200)
+    plot.title = 'Very Small Line Chart 200px'
+    @datasets.each do |data|
+      plot.data(data[0], data[1])
+    end
+    plot.write('spec/reference_images/line_very_small_test.png')
+  end
+
+  it 'test_should_not_hang_with_0_0_100' do
```
Comment (mohawkjohn) [inline spec/common_scripting_layer/line_spec.rb]: What's with the underscores?

File: spec/common_scripting_layer/stacked_bar_spec.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    plot.labels = {
+      0 => '5/6',
+      1 => '5/15',
+      2 => '5/24',
+      3 => '5/30'
+    }
+    @datasets.each do |data|
+      plot.data(data[0], data[1])
+    end
+    plot.write('spec/reference_images/stacked_bar_test_1.png')
```
Comment (mohawkjohn) [inline spec/common_scripting_layer/stacked_bar_spec.rb]: Where is the comparison code for these?

File: spec/common_scripting_layer/pie_spec.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    plot.data(:Arafat, [41])
+    plot.data(:Pranav, [42])
+    #    plot.data(:Grouch, [40])
+    #    plot.data(:Snuffleupagus, [43])
+
+    plot.write('test/output/pie_nearly_equal.png')
+  end
+
+  it 'pie graph equal' do
```
Comment (mohawkjohn) [inline spec/common_scripting_layer/pie_spec.rb]: So, these are really supposed to be written in grammatical English. That's the whole point of this weird `it` notation. These should be formulated as sentences, and if you can't formulate them that way, they may not be good specs (and certainly aren't clear to readers).

`it 'should produce a pie graph with equal proportions' do`, for example.

Comment (Arafatk) [reply spec/common_scripting_layer/pie_spec.rb]: Okay. Will change all of these don't worry about it.

File: spec/common_scripting_layer/histogram_spec.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    plot = Rubyplot::Histogram.new(600)
+    plot.title = 'Random Histogram Numbers'
+    plot.marker_count = 8
+    plot.data('data', [5, 12, 9, 6, 7])
+    plot.write('spec/reference_images/histogram_test_1.png')
+
+    expect(compare_with_reference?('histogram.png', 'histogram_test_1.png', 10)).to eq(true)
+  end
+
+  it 'Fails to match with the reference image' do
```
Comment (mohawkjohn) [inline spec/common_scripting_layer/histogram_spec.rb]: Okay, so this one is grammatically correct (except for the capitalization), but doesn't actually say *why* it should fail to match with the reference image. Does this demonstrate a false comparison?

File: spec/common_scripting_layer/error_bars_spec.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
+describe 'Error Bars!' do
```
Comment (mohawkjohn) [inline spec/common_scripting_layer/error_bars_spec.rb]: Exclamation point?

File: spec/common_scripting_layer/error_bars_spec.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  it 'Creates an error bars plot' do
+    plot = Rubyplot::ErrorBars.new
+    plot.data(@open, @high, @low, @close)
+    plot.write('spec/reference_images/error_bars_test_1.png')
+
+    expect(compare_with_reference?('error_bars.png', \
+                                   'error_bars_test_1.png', 10)).to eq(true)
+  end
+
+  it 'creates a candle plot with blue positive and black negative color' do
```
Comment (mohawkjohn) [inline spec/common_scripting_layer/error_bars_spec.rb]: This is a good example of how a spec title should be written.

File: spec/common_scripting_layer/dot_spec.rb [hunk 7]
```diff
@@ -0,0 +1,102 @@
+require 'spec_helper'
+describe 'Graph' do
```
Comment (mohawkjohn) [inline spec/common_scripting_layer/dot_spec.rb]: You can also do `describe :Graph do`

File: spec/common_scripting_layer/histogram_spec.rb (hunk 4 above)
Comment (Arafatk) [reply spec/common_scripting_layer/histogram_spec.rb]: Here is the difference look closely.  

1. plot.data('data', [5, 12, 9, 6, 7])
2. plot.data('data', [5, 12, 9, 6, 6])

This difference is a very subtle difference that is used to observe the failure.

**Comment by mohawkjohn** [comment]
I've provided some comments which are by no means exhaustive, but I have another question which supersedes those comments. None of these specs seem to reference the matplotlib graph gallery. Why is that? Isn't that part of what you were going to be working on?

File: spec/common_scripting_layer/histogram_spec.rb (hunk 4 above)
Comment (mohawkjohn) [reply spec/common_scripting_layer/histogram_spec.rb]: You are misunderstanding my critique. Your spec titles should _say_ what the difference is. People shouldn't have to search through your spec to find it.

**Comment by Arafatk** [comment]
Hi John,
That's true.
I do have references to the ideal images from Matlplot lib in the google
doc ->
https://docs.google.com/document/d/1PwlxewwKgw7yh-u1y5iJ2f3HPJ2htmlRXf_JCSE7Mj4/edit?usp=sharing
I am not exactly sure what you mean but I am assuming you meant that I
should add comments in the specs to the reference images.
Right?

Regards,
Arafat

**Comment by mohawkjohn** [comment]
You should *always* cite sources in your code and your writing. This is a cardinal rule of science.

* Is this code designed to recapitulate something I saw somewhere else?
* Is this code borrowed from somewhere else?
* Did I get the idea for this code from somewhere else?

You don't just cite sources because it's important to do from an intellectual property standpoint. You do it because it makes your code more maintainable. Anyone who is reading it can go and look to see how you came up with an idea without having to reverse engineer your whole project.

So if you're generating images from the matplotlib gallery, you need to include links to show which ones you're generating.

**Comment by Arafatk** [comment]
Okay, I will always take care of this from now on.

**Comment by Arafatk** [comment]
I mean, I think this is important but don't you think that all
citations/sources
should generally belong to the main class docs rather than specs. In this
case,
in particular, I wouldn't mind keeping them with the specs but I will shift
that comment to the class docs when I finish writing the class.
I hope that's okay.

**Comment by mohawkjohn** [comment]
I'm not sure you're understanding clearly what I'm telling you to cite.

In the libraries for your project (the classes and functions that make it go), you should be citing your sources.

However, in the specs, particularly since we asked you to develop specs based on the matplotlib gallery, you should be linking each spec to a particular gallery entry. This is a citation also.

Since we've been asking you to focus on the specs and you've been working on other things (like bubble plots), I've taken this to mean that you've already finished the specs. Is that correct? If so, you need to do these links now, so we can establish which gallery items you've got represented here.

**Comment by Arafatk** [comment]
I am finishing the specs right now.
Just sending a commit very soon.

**Comment by mohawkjohn** [comment]
Okay. To count as finished, they need to include links to the plots they are based on in the Matplotlib gallery.

File: spec/common_scripting_layer/stacked_bar_spec.rb (hunk 2 above)
Comment (Arafatk) [reply spec/common_scripting_layer/stacked_bar_spec.rb]: Added the line just down below.

**Comment by v0dro** [comment]
@Arafatk : 

Allow me to clarify what John is saying. He wants you to add the sources from the matplotlib gallery from where you have ported specs for the graphs so that when someone sees your specs they can simply refer to the sources and see what has been written.

We don't want you to add the citations directly to the code since we have many different ways of plotting the same thing (multiple backends).

Is this clear enough?

**Comment by v0dro** [comment]
One more suggestion: how about you put the different kinds of plots (`scatter`, `bar`, etc.) into different folders and put the various files for them `scatter_spec.rb` and `scatter_gen.rb` inside the `scatter`folder?

**Comment by v0dro** [comment]
Looks good to me. @mohawkjohn ?

**Comment by v0dro** [comment]
@mohawkjohn ping! Should i merge this?

File: spec/common_scripting_layer/stacked_bar_spec.rb [hunk 8]
```diff
@@ -20,11 +21,12 @@
       plot.data(data[0], data[1])
     end
     plot.write('spec/reference_images/stacked_bar_test_1.png')
+    expect(compare_with_reference?('stacked_bar.png', 'stacked_bar_test_1.png', 10)).to eq(true)
   end
 
-  it 'Tests Many random datapoints' do
+  it 'creates a Stacked Bar graph with preset size from the user' do
```
Comment (mohawkjohn) [inline spec/common_scripting_layer/stacked_bar_spec.rb]: Why the random caps?

File: spec/common_scripting_layer/stack_spec.rb [hunk 9]
```diff
@@ -27,7 +28,10 @@
     plot.write('spec/reference_images/stacked_area_test_1.png')
   end
 
-  it 'Tests Many random datapoints' do
+  it 'creates stacked are graph with preset colors' do
```
Comment (mohawkjohn) [inline spec/common_scripting_layer/stack_spec.rb]: area

File: spec/common_scripting_layer/scatter_spec.rb [hunk 10]
```diff
@@ -29,16 +31,14 @@
     plot.write('spec/reference_images/scatter_many_test.png')
   end
 
-  # Done
-  it 'Test No Title' do
+  it 'makes graph with no title' do
```
Comment (mohawkjohn) [inline spec/common_scripting_layer/scatter_spec.rb]: These ones are better. Good.

**Comment by Arafatk** [comment]
@mohawkjohn Thanks a lot for the comments, I am on it. 
    
I will update you pretty soon.

**Comment by v0dro** [comment]
@Arafatk what's the status on this PR? I'd like to merge it soon as these tests are important for the library.

**Comment by Arafatk** [comment]
@v0dro I think we have added a lot of base specs for now and perhaps there might be minor issues with some conventions here and there but overall, it's okay to merge.         
I will make proper specs with the integration branch....

**Pull request merged**